python/spec/templates/jax/xla_ops.cpp

```diff
@@ -1,169 +1,197 @@
+{%- endfor -%}
+{%- for arg in mod.extra_outputs %}
+    ffi::ResultBuffer<ffi::DataType::F64> {{arg.name}}{% if not loop.last %},{% endif %}
+{%- endfor %}
+) {
+  {# Dimension aliases for readability #}
+  {% for dim in mod.dimensions %}
+  const auto {{dim.name}} = dim{{dim.coords[1]}}({{mod.inputs[dim.coords[0]].name}});
+  {% endfor %}
+  {# Minimal shape checks - rely on driver.hpp order helper #}
+  {% for arg in mod.inputs %}
+  {%- if arg.shape|length == 1 %}
+  if (dim0({{arg.name}}) != {{arg.shape[0]}}) return shape_error("{{mod.name}} shape mismatch");
+  {%- elif arg.shape|length == 2 %}
+  if (dim0({{arg.name}}) != {{arg.shape[0]}} || dim1({{arg.name}}) != {{arg.shape[1]}}) return shape_error("{{mod.name}} shape mismatch");
+  {%- endif %}
+  {% endfor %}
+
+#define FIXED_SIZE_MAP(SIZE)                                                        \
+  {                                                                                 \
+    {%- for arg in mod.inputs %}
+    {%- if arg.shape|length == 1 %}
+    Eigen::Map<const Eigen::VectorXd> {{arg.name}}_({{arg.name}}.typed_data(), {{arg.shape[0]}}, 1); \
+    {%- elif arg.shape|length == 2 and arg.shape[1] == "J" %}
+    Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, SIZE, order<SIZE>::value>> {{arg.name}}_({{arg.name}}.typed_data(), {{arg.shape[0]}}, J); \
     {%- else %}
-    Eigen::Map<{% if not arg.is_output %}const {% endif %}Eigen::VectorXd> {{arg.name}}_({{arg.name}}, {{arg.shape[0]}}, 1); \
-    {%- endif -%}
-    {%- elif arg.shape|length == 2 -%}
-    {%- if arg.shape[1] == "J" %}
-    Eigen::Map<{% if not arg.is_output %}const {% endif %}Eigen::Matrix<double, Eigen::Dynamic, SIZE, order<SIZE>::value>> {{arg.name}}_({{arg.name}}, {{arg.shape[0]}}, J); \
-    {%- endif -%}
-    {%- else -%}
-    {%- if arg.shape[2] == "J" %}
-    Eigen::Map<{% if not arg.is_output %}const {% endif %}Eigen::Matrix<double, Eigen::Dynamic, (SIZE * SIZE), order<(SIZE * SIZE)>::value>> {{arg.name}}_({{arg.name}}, {{arg.shape[0]}}, J * J); \
-    {%- endif -%}
-    {%- endif -%}
-    {% endfor %}
-    {%- if mod.name == "factor" %}
-    Eigen::Index flag = celerite2::core::{{mod.name}}({% for val in mod.inputs + mod.outputs + mod.extra_outputs %}{{val.name}}_{%- if not loop.last %}, {% endif %}{% endfor %}); \
-    if (flag) d_.setZero(); \
+    Eigen::Map<const Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>> {{arg.name}}_({{arg.name}}.typed_data(), {{arg.shape[0]}}, dim1({{arg.name}})); \
+    {%- endif %}
+    {%- endfor %}
+    {%- for arg in mod.outputs %}
+    {%- if arg.shape|length == 1 %}
+    Eigen::Map<Eigen::VectorXd> {{arg.name}}_({{arg.name}}->typed_data(), {{arg.shape[0]}}, 1); \
+    {%- elif arg.shape|length == 2 and arg.shape[1] == "J" %}
+    Eigen::Map<Eigen::Matrix<double, Eigen::Dynamic, SIZE, order<SIZE>::value>> {{arg.name}}_({{arg.name}}->typed_data(), {{arg.shape[0]}}, J); \
     {%- else %}
-    if (nrhs == 1) { \
```

Comment:
This condition seems to be missing from the new implementation. Where did it go?

Reply:
Oops. Looks like I deleted it in the refactor. My bad. One sec.